connectors/stash/src/main/java/com/vmware/connectors/stash/StashController.java

```diff
@@ -0,0 +1,444 @@
+                                .setMinLength(1)
+                                .build()
+                    )
+                    .setType(HttpMethod.POST)
+                    .build()
+                );
+    }
+
+    private String buildActionUrl(final String routingPrefix,
+                                  final StashPullRequest pullRequest,
+                                  final StashAction stashAction) {
+        return String.format(StashConstants.STASH_CLIENT_ACTION_URL,
+                routingPrefix,
+                pullRequest.getProjectKey(),
+                pullRequest.getRepositorySlug(),
+                pullRequest.getPullRequestId(),
+                stashAction.getAction());
+    }
+
+    private CardBody makeCardBody(final JsonDocument stashResponse, final List<String> comments) {
+        final CardBody.Builder cardBody = new CardBody.Builder();
+        cardBody.setDescription(stashResponse.read("$.description"));
+
+        cardBody.addField(makeCardBodyField("stash.repository", stashResponse.read("$.fromRef.repository.slug")));
+        cardBody.addField(makeCardBodyField("stash.author", stashResponse.read("$.author.user.displayName")));
+        cardBody.addField(makeCardBodyField("stash.title", stashResponse.read("$.title")));
+        cardBody.addField(makeCardBodyField("stash.state", stashResponse.read("$.state")));
+        cardBody.addField(makeCardBodyField("stash.open", Boolean.toString(stashResponse.read("$.open"))));
+
+        if (!CollectionUtils.isEmpty(comments)) {
+            cardBody.addField(addCommentField(comments));
+        }
+        return cardBody.build();
+    }
+
+    private CardBodyField addCommentField(final List<String> comments) {
+        final CardBodyField.Builder cardBodyFieldBuilder = new CardBodyField.Builder();
+        cardBodyFieldBuilder.setTitle(this.cardTextAccessor.getMessage(STASH_COMMENTS));
+        cardBodyFieldBuilder.setType(CardBodyFieldType.COMMENT);
+
+        comments.forEach(comment -> {
+            cardBodyFieldBuilder.addContent(ImmutableMap.of("text", comment));
+        });
+        return cardBodyFieldBuilder.build();
+    }
+
+    private CardBodyField makeCardBodyField(final String messageKeyPrefix, final String descriptionArgs) {
+        return new CardBodyField.Builder()
+                .setTitle(cardTextAccessor.getMessage(messageKeyPrefix + ".title"))
+                .setType(CardBodyFieldType.GENERAL)
+                .setDescription(
+                        cardTextAccessor.getMessage(
+                                messageKeyPrefix + ".description",
+                                descriptionArgs
+                        )
+                )
+                .build();
+    }
+
+    private Set<StashPullRequest> convertToStashPullRequest(final Set<String> cardTokens) {
+        final Set<StashPullRequest> pullRequests = new HashSet<>();
+        final Pattern pattern = Pattern.compile(StashConstants.STASH_PR_EMAIL_SUBJECT_REGEX);
```

Comment:
It's unfortunate that we have to do this, but I don't have a better solution in mind.  We could try to look for the "View pull request" link in the email, but that gets ugly fast:

```
<a href=3D"https://stash.air-watch.com/projects/UFO/repos/app-platform-server=
/pull-requests/252" target=3D"_blank" style=3D"color: #3b73af; text-decoration: =
none; border: 0; text-decoration: none; color: #ffffff">View pull request</a=
>
```
What you have is probably our best option right now.

Reply:
Yeah.. Extracting the pull request email subject would be an easier option for the client..